### PR TITLE
feat(go): parse main mod version from build info settings

### PR DESCRIPTION
**Pull request by oatovar**
## Description

* Parse the build info for -ldflags. If included, the flags will be further parsed, and searched for -X sub-flags. These sub-flags are commonly used to set the binary version during build time. A common pattern is to run `go build -ldflags='-X main.version=<semver>'` so that the main package's version variable is replaced with the latest tag. It's not guaranteed that this flag will propogate into the binary. See https://github.com/golang/go/issues/63432 for more info.

* Flag parsing reuses the pflags library that's used by the main Trivy binary. This keeps the implementation concise, and a bit more robust than creating a custom flag parser for -X flag commonly passed to -ldflags.

* Add simple binary to test for validity of ldflags parsing.

* The documentation has been updated to reflect the improved accuracy of the Go binary parser.

## Related issues

- Relates to #1837 
- Relates to #6521 

## Checklist

- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

## Output in CycloneDX

```json
    {
      "bom-ref": "pkg:golang/github.com/aquasecurity/trivy@0.50.1-65-ge80e222e9",
      "type": "library",
      "name": "github.com/aquasecurity/trivy",
      "version": "0.50.1-65-ge80e222e9",
      "purl": "pkg:golang/github.com/aquasecurity/trivy@0.50.1-65-ge80e222e9",
      "properties": [
        {
          "name": "aquasecurity:trivy:LayerDiffID",
          "value": "sha256:41f456757360f81ad1905937cc923e2d07edb9f590e1de739e18a56543b905c1"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "gobinary"
        }
      ]
    }
```

### CODE REVIEW

**Comment by oatovar** [comment]
I used [a test project](https://github.com/oatovar/go-ldflag-version-stamped-test) to test the new parsing ability with test data. I wasn't sure how to generate the test data otherwise. Please let me know if there's something that I've missed about this.

File: pkg/dependency/parser/golang/binary/parse.go [hunk 1]
```diff
@@ -48,22 +51,25 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
-			Version:      p.checkVersion(info.Main.Path, info.Main.Version),
+			Version:      cmp.Or(p.checkVersion(info.Main.Path, info.Main.Version), p.parseLDFlags(info.Main.Path, ldflags)),
 			Relationship: types.RelationshipRoot,
 		},
 		{
 			// Add the Go version used to build this binary.
 			Name:         "stdlib",
 			Version:      strings.TrimPrefix(info.GoVersion, "go"),
-			Relationship: types.RelationshipDirect, // Considered a direct dependency as the main module depends on the standard packages.
+			Relationship: types.RelationshipDirect, // Considered a direct dependency as the main
```
Comment (oatovar) [inline pkg/dependency/parser/golang/binary/parse.go:72]:
```suggestion
			Relationship: types.RelationshipDirect, // Considered a direct dependency as the main module depends on the standard packages.
```

**Comment by oatovar** [comment]
Fixed the unit test that changed after the merge conflict. 😢 I forgot to run `mage test:unit` before pushing that resolution, sorry.

File: pkg/dependency/parser/golang/binary/parse.go [hunk 2]
```diff
@@ -93,8 +99,69 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+		xx = strings.Trim(xx, `'"`)
+		key, val, found := strings.Cut(xx, "=")
+		if !found {
+			p.logger.Debug("Unable to parse an -ldflags -X flag value", "got", xx)
+			continue
+		}
+
+		key = strings.ToLower(key)
+		// The check for a 'ver' prefix enables the parser to pick up Trivy's own version value that's set.
+		if strings.HasSuffix(key, "version") || strings.HasSuffix(key, "ver") {
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:160]: What if validating semver?

```suggestion
		if (strings.HasSuffix(key, "version") || strings.HasSuffix(key, "ver")) && semver.IsValid(value) {
```

File: pkg/dependency/parser/golang/binary/parse.go [hunk 3]
```diff
@@ -93,8 +99,69 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+	fset := pflag.NewFlagSet("ldflags", pflag.ContinueOnError)
+	// This prevents the flag set from erroring out if other flags were provided.
+	// This helps keep the implementation small, so that only the -X flag is needed.
+	fset.ParseErrorsWhitelist.UnknownFlags = true
+	// The shorthand name is needed here because setting the full name
+	// to `X` will cause the flag set to look for `--X` instead of `-X`.
+	// The flag can also be set multiple times, so a string slice is needed
+	// to handle that edge case.
+	var x []string
+	fset.StringSliceVarP(&x, "", "X", nil, `Set the value of the string variable in importpath named name to value.
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:131]: I don't think we need the original usage here.

File: pkg/dependency/parser/golang/binary/parse.go [hunk 4]
```diff
@@ -93,8 +99,69 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+		// -X 'main.version=1.0.0'
+		// -X='main.version=1.0.0'
+		// -X="main.version=1.0.0"
+		// -X "main.version=1.0.0"
+		xx = strings.Trim(xx, `'"`)
+		key, val, found := strings.Cut(xx, "=")
+		if !found {
+			p.logger.Debug("Unable to parse an -ldflags -X flag value", "got", xx)
+			continue
+		}
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:156]: `StringToStringVarP` possibly simplifies parsing.

```suggestion
	var x map[string]string
	fset.StringToStringVarP(&x, "", "X", nil, "")
	if err := fset.Parse(flags); err != nil {
		p.logger.Error("Could not parse -ldflags found in build info", "err", err)
		return ""
	}

	for key, value := range x {
		key = strings.ToLower(key)
		// The check for a 'ver' prefix enables the parser to pick up Trivy's own version value that's set.
		if (strings.HasSuffix(key, "version") || strings.HasSuffix(key, "ver")) && semver.IsValid(value) {
			return value
		}
	}```

File: pkg/dependency/parser/golang/binary/parse.go [hunk 5]
```diff
@@ -93,8 +99,69 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+
+		return strings.Fields(setting.Value)
+	}
+	return nil
+}
+
+// parseLDFlags attempts to parse the binary's version from any `-ldflags` passed to `go build` at build time.
+func (p *Parser) parseLDFlags(name string, flags []string) string {
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:120]: The input is so varied that it is worth exporting this method and writing tests for it.

```suggestion
func (p *Parser) ParseLDFlags(name string, flags []string) string {
```

```
func TestParser_ParseLDFlags(t *testing.T) {
	type args struct {
		name  string
		flags []string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		{
			name: "normal",
			args: args{
				name: "github.com/aquasecurity/trivy",
				flags: []string{
					"-s",
					"-w",
					"-X=foo=bar",
					"-X='github.com/aquasecurity/trivy/pkg/version.ver=0.50.1'",
				},
			},
			want: "0.50.1",
		},
		{
			name: "empty",
			args: args{
				name:  "github.com/aquasecurity/test",
				flags: []string{},
			},
			want: "",
		},
                ... (more tests) ...
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := binary.NewParser().(*binary.Parser)
			assert.Equal(t, tt.want, p.ParseLDFlags(tt.args.name, tt.args.flags))
		})
	}
}
```

File: pkg/dependency/parser/golang/binary/parse.go [hunk 6]
```diff
@@ -93,8 +99,69 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+	log.Debug("Parsing dependency's build info settings", "dependency", name, "-ldflags", flags)
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:121]:
```suggestion
	p.logger.Debug("Parsing dependency's build info settings", "dependency", name, "-ldflags", flags)
```

**Review by knqyf263 — COMMENTED** [review]
Thanks for your great contribution! Looks almost good. I left some comments.

We'd love to include it in [v0.51.0](https://github.com/aquasecurity/trivy/milestone/33), which will be released this week. Please let me know if you don't have time. We'll add small tweaks and wrap it up.

**Comment by oatovar** [comment]
@knqyf263 Thank you for the review. Having use the pflag library in the past, I had overlooked the `StringToStringVarP` functionality - it's very powerful! I've applied your suggestions, and I think the PR is ready for another review.

**Comment by knqyf263** [comment]
Great! I'll take another look today.

File: pkg/dependency/parser/golang/binary/parse.go [hunk 7]
```diff
@@ -93,8 +100,71 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+
+func isValidXKey(key string) bool {
+	key = strings.ToLower(key)
+	// The check for a 'ver' prefix enables the parser to pick up Trivy's own version value that's set.
+	return strings.HasSuffix(key, "version") || strings.HasSuffix(key, "ver")
+}
+
+func isValidSemVer(ver string) bool {
+	// semver.IsValid strictly checks for the v prefix so prepending 'v'
+	// here and checking validaty again increases the chances that we
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:167]: nit

```suggestion
	// here and checking validity again increases the chances that we
```

File: pkg/dependency/parser/golang/binary/parse.go [hunk 8]
```diff
@@ -93,8 +100,71 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+		// have been provided: Ex:
+		//
+		// -X main.version=1.0.0
+		// -X=main.version=1.0.0
+		// -X 'main.version=1.0.0'
+		// -X='main.version=1.0.0'
+		// -X="main.version=1.0.0"
+		// -X "main.version=1.0.0"
+		key = strings.TrimLeft(key, `'"`)
+		val = strings.TrimRight(val, `'"`)
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:149]: nit: I believe double quotes are handled by pflag. Please correct me if I'm wrong.

```suggestion
		key = strings.TrimLeft(key, `'`)
		val = strings.TrimRight(val, `'`)
```

File: pkg/dependency/parser/golang/binary/parse.go [hunk 9]
```diff
@@ -93,8 +100,71 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+	// This helps keep the implementation small, so that only the -X flag is needed.
+	fset.ParseErrorsWhitelist.UnknownFlags = true
+	// The shorthand name is needed here because setting the full name
+	// to `X` will cause the flag set to look for `--X` instead of `-X`.
+	// The flag can also be set multiple times, so a string slice is needed
+	// to handle that edge case.
+	var x map[string]string
+	fset.StringToStringVarP(&x, "", "X", nil, "")
+	if err := fset.Parse(flags); err != nil {
+		p.logger.Error("Could not parse -ldflags found in build info", "err", err)
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse.go:134]: nit

```suggestion
		p.logger.Error("Could not parse -ldflags found in build info", log.Err(err))
```

File: pkg/dependency/parser/golang/binary/parse_test.go [hunk 10]
```diff
@@ -122,3 +138,108 @@ func TestParse(t *testing.T) {
...
+					"-s",
+					"-w",
+					"-X=foo=bar",
+					"-X='github.com/aquasecurity/trivy/pkg/version.Ver=v0.50.1'",
+				},
+			},
+			want: "v0.50.1",
+		},
+		{
+			name: "with double quotted flag",
```
Comment (knqyf263) [inline pkg/dependency/parser/golang/binary/parse_test.go:205]: nit: I'm not sure, but my editor is complaining.

```suggestion
			name: "with double quoted flag",
```

**Review by knqyf263 — COMMENTED** [review]
Looks good! I left very small comments.

File: pkg/dependency/parser/golang/binary/parse.go (hunk 8 above)
Comment (oatovar) [reply pkg/dependency/parser/golang/binary/parse.go:149]: You are correct! It's only single quotes that aren't handled.

**Comment by oatovar** [comment]
Thank you for catching those errors 😄 I've applied them all.

**Review by knqyf263 — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**